### PR TITLE
8335480: Only deoptimize threads if needed when closing shared arena

### PR DESCRIPTION
**Pull request by JornVernee**
This PR limits the number of cases in which we deoptimize frames when closing a shared Arena. The initial intent of this was to improve the performance of shared arena closure in cases where a lot of threads are accessing and closing shared arenas at the same time (see attached benchmark), but unfortunately even disabling deoptimization altogether does not have a great effect on that benchmark.

Nevertheless, I think the extra logging/testing/benchmark code, and comments I've written, together with reducing the number of cases where we deoptimize, which makes it clearer exactly why we need to deoptimize in the first place, will be useful going forward. So, I've a create this PR out of them.

In this PR:
- I've separate the stack walking code (`for_scope_method`) from the code that checks for a reference to the arena being closed (`is_accessing_session`), and added logging code to the former. That also required changing vframe code to accept an `ouputStream*` rather than always printing to `tty`.
- Added a new test (`TestConcurrentClose`), that tries to close many shared arenas at the same time, in order to stress that use case.
- Added a new benchmark (`ConcurrentClose`), that stresses the cases where many threads are accessing and closing shared arenas.

I've done several benchmark runs with different amounts of threads. The confined case stays much more consistent, while the shared cases balloons up in time spent quickly when there are more than 4 threads:

```
Benchmark                     Threads   Mode  Cnt     Score     Error  Units
ConcurrentClose.sharedAccess       32   avgt   10  9017.397 ± 202.870  us/op
ConcurrentClose.sharedAccess       24   avgt   10  5178.214 ± 164.922  us/op
ConcurrentClose.sharedAccess       16   avgt   10  2224.420 ± 165.754  us/op
ConcurrentClose.sharedAccess        8   avgt   10   593.828 ±   8.321  us/op
ConcurrentClose.sharedAccess        7   avgt   10   470.700 ±  22.511  us/op
ConcurrentClose.sharedAccess        6   avgt   10   386.697 ±  59.170  us/op
ConcurrentClose.sharedAccess        5   avgt   10   291.157 ±   7.023  us/op
ConcurrentClose.sharedAccess        4   avgt   10   209.178 ±   5.802  us/op
ConcurrentClose.sharedAccess        1   avgt   10    52.042 ±   0.630  us/op
ConcurrentClose.confinedAccess     32   avgt   10    25.517 ±   1.069  us/op
ConcurrentClose.confinedAccess      1   avgt   10    12.398 ±   0.098  us/op
```

(I manually added the `Threads` collumn btw)

Testing: tier 1-4

### CODE REVIEW

**Comment by JornVernee** [comment]
>  This could be narrowed down further by tracking for each compiled method if it has an (inlined) call to an `@Scoped` method, but I've left that out for now.

I decided to add this to the PR as well for completeness, so that we don't go and deoptimize frames that are not using scoped accesses at all.

**Comment by forax** [comment]
Nice work !

Thinking a bit about how to improve the benchmark and given the semantics of Arena.close(), there is a trick that you can use. There are two kinds of memory segments, the one that only visible from Java and the one that are visible not only from Java. By example, a memory segment created from a mmap or a memory segment with an address sent to a C code are visible from outside Java, for those, you have no choice but wait in Arena.close() until all threads have answered to the handshakes. For all the other memory segments, because they are only visible from Java, their memory can be reclaimed asynchronously, i.e. the last thread of the handshakes can free the corresponding memory segments, so the thread that call Arena.close() is free to run even if the memory is not yet reclaimed.

From my armchair, that seems a awful lot of engeneering so it may not worth it, but now you know :)

**Comment by JornVernee** [comment]
That is something we considered in the past as well (I think Maurizio even had a prototype at some point). The issue is that close should be deterministic. i.e. after the call to `close()` returns, all memory should be freed. That is an essential property for applications that have most of their virtual address space tied up, and then want to release and immediately re-use it. If it's not important that memory is freed deterministically, but memory should still be accessible from multiple threads, an automatic arena might be a better choice.

**Review by fisk — APPROVED** [review]
Looks good.

**Comment by fisk** [comment]
@dougxc might want to have a look at Graal support for this one.

**Comment by JornVernee** [comment]
> @dougxc might want to have a look at Graal support for this one.

Yes, I conservatively implemented `has_scoped_access()` for Graal. It won't regress anything, but there's still an opportunity for improvement.

**Comment by forax** [comment]
Knowing that all the segments are freed during close() is something you may want.
But having the execution time of close() be linear with the number of threads is also problematic. Maybe, it means that we need another kind of Arena that works like shared() but allow the freed to be done asynchronously (ofSharedAsyncFree ?).

Note that the semantics of ofSharedAsyncFree() is different from ofAuto(), ofAuto() relies on the GC to free a segment so the delay before a segment is freed is not time bounded if the application has enough memory, the memory of the segment may never be reclaimed. With ofSharedAsyncFree(), the segments are freed by the last thread, so while this mechanism is not deterministic, it is time bounded.

**Comment by uschindler** [comment]
Hi Jorn,

Many thanks for working on this! 

I have one problem with the benchmark: I think it is not measuring the whole setup in a way that is our workload: The basic problem is that we don't want to deoptimize threads which are not related to MemorySegments. So basically, the throughput of those threads should not be affected. For threads currently in a memory-segment read it should have a bit of effect, but it should recover fast.

The given benchmark somehow only measures the following: It starts many threads; in each it opens a shared memory segment, does some work and closes it. So it measures the throughput of the whole "create shared/work on it/close shared" workload. Actually the problems we see in Lucene are more that we have many threads working on shared memory segments or on other tasks not related to memory segments at all, while a few threads are concurrently closing and opening new arenas. With more threads concurrently closing the arenas, also the throughput on other threads degrades.

So IMHO, the benchamrk should be improved to have a few threads (configurable) that open/close memory segments and a list of other threads that do other work and finally a list of threads reading from the memory segments opened by first thread. The testcase you wrote is more fitting the above workload. Maybe the benchmark should be setup more like the test. If you have a benchmark with that workload it should better show an improvement.

The current benchmark has the problem that it measures the whole open/work/close on shared sgements. And slosing a shared segment is always heavy, because it has to trigger and wait for the thread-local handshake.

Why is the test preventing inlining of the inner read method?

I may be able to benchmark a Lucene workload with a custom JDK build next week. It might be an idea to use the wrong DaCapoBenchmark (downgrade to older version before it has fixed https://github.com/dacapobench/dacapobench/issues/264 , specifically https://github.com/dacapobench/dacapobench/commit/76588b28d516ae19f51a80e7287d404385a2c146).

Uwe

**Comment by uschindler** [comment]
> Knowing that all the segments are freed during close() is something you may want. But having the execution time of close() be linear with the number of threads is also problematic. Maybe, it means that we need another kind of Arena that works like shared() but allow the freed to be done asynchronously (ofSharedAsyncFree ?).
> 
> Note that the semantics of ofSharedAsyncFree() is different from ofAuto(), ofAuto() relies on the GC to free a segment so the delay before a segment is freed is not time bounded if the application has enough memory, the memory of the segment may never be reclaimed. With ofSharedAsyncFree(), the segments are freed by the last thread, so while this mechanism is not deterministic, it is time bounded.

That's a great suggestion! In our case we just want the index files open as soon as possible, but not on next GC (which will be horrible and brings us back into the times of DirectByteBuffer). The problem with GC is that the Arena/MemorySegments and so on are tiny objects which will live for very long time, especially when they were used for quite some time (like an index segment of an Lucene index).

Of course for testing purposes in Lucene we could use `ofShared()` (to make sure all mmapped files are freeed, especially on Windows as soon as index is close), but in production environments we could offer the option to use delayed close to improve throughput.

Uwe

File: src/hotspot/share/prims/scopedMemoryAccess.cpp [hunk 1]
```diff
@@ -104,38 +146,62 @@ class CloseScopedMemoryClosure : public HandshakeClosure {
...
+        // liveness check (from @Scoped method)
+        // for (...) {
+        //    for (...) { // strip-mining inner loop
+        //        memory access (from @Scoped method)
+        //    }
+        //    safepoint <-- STOPPED HERE
+        // }
+        //
+        // The safepoint at which we're stopped may be in between the liveness check
+        // and actual memory access, but is itself 'outside' of @Scoped code
```
Comment (dougxc) [inline src/hotspot/share/prims/scopedMemoryAccess.cpp:179]: what is `@Scoped code`? I don't see that annotation mentioned here: https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/ScopedValue.html

Comment (uschindler) [reply src/hotspot/share/prims/scopedMemoryAccess.cpp:179]: This is the whole magic around the shared arena. It is not public API and internal to Hotspot/VM:
- https://github.com/openjdk/jdk/blob/a96de6d8d273d75a6500e10ed06faab9955f893b/src/java.base/share/classes/jdk/internal/misc/X-ScopedMemoryAccess.java.template#L117-L119
- https://github.com/openjdk/jdk/blob/a96de6d8d273d75a6500e10ed06faab9955f893b/src/hotspot/share/prims/scopedMemoryAccess.cpp#L143-L149

Comment (AlanBateman) [reply src/hotspot/share/prims/scopedMemoryAccess.cpp:179]: > what is `@Scoped code`? I don't see that annotation mentioned here: https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/ScopedValue.html

This is nothing to do with scoped values, instead this is an annotation declared in jdk.internal.misc.ScopedMemoryAccess that is known to the VM.

File: src/hotspot/share/jvmci/jvmciRuntime.cpp [hunk 2]
```diff
@@ -2183,6 +2183,7 @@ JVMCI::CodeInstallResult JVMCIRuntime::register_method(JVMCIEnv* JVMCIENV,
         nm->set_has_unsafe_access(has_unsafe_access);
         nm->set_has_wide_vectors(has_wide_vector);
         nm->set_has_monitors(has_monitors);
+        nm->set_has_scoped_access(true); // conservative
```
Comment (dougxc) [inline src/hotspot/share/jvmci/jvmciRuntime.cpp:2186]: What does "conservative" imply here? That is, what performance penalty will be incurred for Graal compiled code until it completely supports this "scoped access" bit?

File: src/hotspot/share/prims/scopedMemoryAccess.cpp (hunk 1 above)
Comment (uschindler) [reply src/hotspot/share/prims/scopedMemoryAccess.cpp:179]: Basically if the VM is inside a @Scoped method and it starts a thread-local handshake, it will deoptimize top-most frame of all those threads so they can do the "isAlive" check.

File: test/micro/org/openjdk/bench/java/lang/foreign/ConcurrentClose.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import java.lang.foreign.Arena;
+import java.lang.foreign.MemorySegment;
+import java.lang.invoke.MethodHandle;
+import java.util.concurrent.TimeUnit;
+
+import org.openjdk.jmh.annotations.*;
+
+import static java.lang.foreign.ValueLayout.*;
+
+@BenchmarkMode(Mode.AverageTime)
```
[the remark this answers is not in the document]
Comment (uschindler) [reply test/micro/org/openjdk/bench/java/lang/foreign/ConcurrentClose.java:34]: It should also run threads not doing any scoped accesses to verify that other threads are not affected.

**Comment by uschindler** [comment]
> > I have one problem with the benchmark: I think it is not measuring the whole setup in a way that is our workload: The basic problem is that we don't want to deoptimize threads which are not related to MemorySegments. So basically, the throughput of those threads should not be affected. For threads currently in a memory-segment read it should have a bit of effect, but it should recover fast.
> 
> IMHO there is a bit of confusion in this discussion. When we say that a shared arena close operation is slow, we might mean one of two things:
> 
> 1. calling the `close()` method itself is slow (this is what the benchmark effectively measures)
> 2. throughput of unrelated threads is affected (I think this is what Lucene is seeing)
> 
> Addressing (2) than (1) (in the sense that, if you sign up for a shared arena close, you know it's going to be deterministic, but expensive, as the javadoc itself admits).

I fully agree, we mixed two different approaches. The problem is that the benchmark measures both, 1 and 2 per thread. To see an effect of this change, the benchmark should have 3 types of threads: One only closing arenas, another set that consumes scoped memory and a third group doing totally unrelated stuff.

> For this reason, I'm unsure about some of the "delaying tactics" I see mentioned here: if we delay the underlying "free"/"unmap" operation, this is only going to affect (1). You still need some global operation (e.g. handshake) to make sure all threads agree on the segment state. Moving the cost of the free/unmap from one place to another is not really going to do much for (2).

This is indeed unrelated. It is just an idea I also thorught of. In Apache Lucene we are mostly interested to close the shared arena as soon as possible. We don't need to make sure it is closed after the "close" call finished (we don't care), but we can't wait until GC closes the arena possibly after hours or even days. The reason for the latter is that the Arena is a small, long-living instance and GC does not want to free it, as there is no pressure.

So basically for us it would be best to trigger the close and then do other stuff.

Of course we can do that in a separate thread (this is my idea how to improve the closes in lucene). The only problem is that Lucene does not have own threadpools, so this would be responsibility of the caller to possibly close our indexes in a separate thread (and a single one only).

**Comment by mcimadamore** [comment]
> One only closing arenas, another set that consumes scoped memory and a third group doing totally unrelated stuff.

Exactly. My general feeling is that the cost of handshaking a thread dominates everything else, so doing improvements around e.g. avoiding unnecessary deoptimization (as in this PR) is not going to help much, even for threads doing unrelated stuff, but I'd be happy to be proved wrong.

**Comment by mcimadamore** [comment]
> avoiding unnecessary deoptimization (as in this PR) is not going to help much,

What would definitively help is to somehow reduce the number of threads to handshake when calling close - e.g. have an arena that is shared but only to a *group* of thread. We can do that easily using structured concurrency. But for unstructured code there's not a lot that can be done, as there's no way for the runtime to guess which threads can access segments created by a given arena.

**Comment by uschindler** [comment]
@mcimadamore: FYI, at the moment we are working on grouping mmapped files together (by their index segment file pattern) and use the same arena for multiple index files. Because those are closed together we can use a refcounted aproach. All files of a group (the index segment name) share the same arena and this one is closed after last file in group is closed: https://github.com/apache/lucene/pull/13570

File: src/hotspot/share/jvmci/jvmciRuntime.cpp (hunk 2 above)
Comment (JornVernee) [reply src/hotspot/share/jvmci/jvmciRuntime.cpp:2186]: It means we will always deoptimize a top-most frame of any thread, when closing a shared arena, and it is compiled by Graal. (This is a one-off deoptimization though. The compiled code is not thrown away). It essentially matches the current behavior before this PR.

**Comment by forax** [comment]
> This is what I was thinking of as well. close() on a shared arena can be called by any thread, so it would be possible to have an executor service with 1-n threads that is dedicated to closing memory.

This delays both the closing of the Arena and the freeing of the segments, so bugs may be not discovered if the arena is accessed in between the time the thread pool is notified and the time the close() is effectively called.

And you loose the structured part of the API, you can not use a try-with-resources anymore. I think that part can be fixed using a wrapper on top of Arena.ofShared().

**Comment by JornVernee** [comment]
I've update the benchmark to run with 3 separate threads: 1 thread that is just creating and closing shared arenas in a loop, 1 that is accessing memory using the FFM API, and 1 that is accessing a `byte[]`.

Current:
```
Benchmark                                        Mode  Cnt   Score    Error  Units
ConcurrentClose.sharedClose                      avgt   10  50.093 ±  6.200  us/op
ConcurrentClose.sharedClose:closing              avgt   10  46.269 ±  0.786  us/op
ConcurrentClose.sharedClose:memorySegmentAccess  avgt   10  98.072 ± 19.061  us/op
ConcurrentClose.sharedClose:otherAccess          avgt   10   5.938 ±  0.058  us/op
```

I do see a pretty big difference on the memory segment accessing thread when I remove deoptimization altogether:

```
Benchmark                                        Mode  Cnt   Score   Error  Units
ConcurrentClose.sharedClose                      avgt   10  22.664 ± 0.409  us/op
ConcurrentClose.sharedClose:closing              avgt   10  45.351 ± 1.554  us/op
ConcurrentClose.sharedClose:memorySegmentAccess  avgt   10  16.671 ± 0.251  us/op
ConcurrentClose.sharedClose:otherAccess          avgt   10   5.969 ± 0.089  us/op
```

When I remove the `has_scoped_access()` check before the deopt, I expect the `otherAccess` thread to be affected, but the effect isn't nearly as big as with the FFM thread. I think this is likely due to the `otherAccess` benchmark being less sensitive to optimization (i.e. it already runs fairly fast in the interpreter). I also tried using `MethodHandles::arrayElementGetter` for the access, but the numbers I got were pretty much the same:

```
Benchmark                                        Mode  Cnt    Score   Error  Units
ConcurrentClose.sharedClose                      avgt   10   52.745 ± 1.071  us/op
ConcurrentClose.sharedClose:closing              avgt   10   46.670 ± 0.453  us/op
ConcurrentClose.sharedClose:memorySegmentAccess  avgt   10  102.663 ± 3.430  us/op
ConcurrentClose.sharedClose:otherAccess          avgt   10    8.901 ± 0.109  us/op
```

I think, to really test the effect of the `has_scoped_access` check, we need to look at a more realistic scenario.

**Comment by mcimadamore** [comment]
> I've update the benchmark to run with 3 separate threads: 1 thread that is just creating and closing shared arenas in a loop, 1 that is accessing memory using the FFM API, and 1 that is accessing a `byte[]`.
> 
> Current:
> 
> ```
> Benchmark                                        Mode  Cnt   Score    Error  Units
> ConcurrentClose.sharedClose                      avgt   10  50.093 ±  6.200  us/op
> ConcurrentClose.sharedClose:closing              avgt   10  46.269 ±  0.786  us/op
> ConcurrentClose.sharedClose:memorySegmentAccess  avgt   10  98.072 ± 19.061  us/op
> ConcurrentClose.sharedClose:otherAccess          avgt   10   5.938 ±  0.058  us/op
> ```
> 
> I do see a pretty big difference on the memory segment accessing thread when I remove deoptimization altogether:
> 
> ```
> Benchmark                                        Mode  Cnt   Score   Error  Units
> ConcurrentClose.sharedClose                      avgt   10  22.664 ± 0.409  us/op
> ConcurrentClose.sharedClose:closing              avgt   10  45.351 ± 1.554  us/op
> ConcurrentClose.sharedClose:memorySegmentAccess  avgt   10  16.671 ± 0.251  us/op
> ConcurrentClose.sharedClose:otherAccess          avgt   10   5.969 ± 0.089  us/op
> ```
> 
> When I remove the `has_scoped_access()` check before the deopt, I expect the `otherAccess` thread to be affected, but the effect isn't nearly as big as with the FFM thread. I think this is likely due to the `otherAccess` benchmark being less sensitive to optimization (i.e. it already runs fairly fast in the interpreter). I also tried using `MethodHandles::arrayElementGetter` for the access, but the numbers I got were pretty much the same:
> 
> ```
> Benchmark                                        Mode  Cnt    Score   Error  Units
> ConcurrentClose.sharedClose                      avgt   10   52.745 ± 1.071  us/op
> ConcurrentClose.sharedClose:closing              avgt   10   46.670 ± 0.453  us/op
> ConcurrentClose.sharedClose:memorySegmentAccess  avgt   10  102.663 ± 3.430  us/op
> ConcurrentClose.sharedClose:otherAccess          avgt   10    8.901 ± 0.109  us/op
> ```
> 
> I think, to really test the effect of the `has_scoped_access` check, we need to look at a more realistic scenario.

Interesting benchmark. What is the baseline here? E.g. can we also compare against same benchmark that is using a confined arena to do the closing?

**Comment by mcimadamore** [comment]
> When I remove the `has_scoped_access()` check before the deopt, I expect the `otherAccess` thread to be affected, but the effect isn't nearly as big as with the FFM thread. I think this is likely due to the `otherAccess` benchmark being less sensitive to optimization (i.e. it already runs fairly fast in the interpreter). I also tried using `MethodHandles::arrayElementGetter` for the access, but the numbers I got were pretty much the same:

To put this into perspective, once the underlying bug with reachability fences is addressed, then we should see the numbers for this benchmark align with the ones where you removed deopt completely (as we won't deopt threads that don't have the target arena in their oopmap)

**Comment by mcimadamore** [comment]
> I also tried using `MethodHandles::arrayElementGetter` for the access, but the numbers I got were pretty much the same:

This is quite strange, as the code involved should be quite similar to those with memory segments (e.g. you go through a method handle pointing to some helper class). I would have said this would have provided a fairly good comparison.

**Comment by JornVernee** [comment]
This is the baseline if I change `closing` to use a confined arena:

```
Benchmark                                        Mode  Cnt   Score    Error  Units
ConcurrentClose.sharedClose                      avgt   10   8.089 ±  0.006  us/op
ConcurrentClose.sharedClose:closing              avgt   10   0.001 ±  0.001  us/op
ConcurrentClose.sharedClose:memorySegmentAccess  avgt   10  20.046 ±  0.019  us/op
ConcurrentClose.sharedClose:otherAccess          avgt   10   4.220 ±  0.002  us/op
```

**Comment by JornVernee** [comment]
> Ah! I had `arrayElementVarHandle` in mind - maybe you can try that?

Even with `arrayElementVarHandle` it's about the same

**Comment by mcimadamore** [comment]
> This is the baseline if I change `closing` to use a confined arena:
> 
> ```
> Benchmark                                        Mode  Cnt   Score    Error  Units
> ConcurrentClose.sharedClose                      avgt   10   8.089 ±  0.006  us/op
> ConcurrentClose.sharedClose:closing              avgt   10   0.001 ±  0.001  us/op
> ConcurrentClose.sharedClose:memorySegmentAccess  avgt   10  20.046 ±  0.019  us/op
> ConcurrentClose.sharedClose:otherAccess          avgt   10   4.220 ±  0.002  us/op
> ```

This is promising. Effectively, once all the issues surrounding reachability fences will be addressed, we should be able to achieve numbers similar to above even in the case of shared close. The only thing being slower in that case would be the closing thread itself.

**Comment by mcimadamore** [comment]
> Even with `arrayElementVarHandle` it's about the same

This is very odd, and I don't have a good explanation as to why that is the case. What does the baseline (confined arena) look like for `arrayElementVarHandle` ?

**Comment by JornVernee** [comment]
> > Even with `arrayElementVarHandle` it's about the same
> 
> This is very odd, and I don't have a good explanation as to why that is the case. What does the baseline (confined arena) look like for `arrayElementVarHandle` ?

Pretty much exactly the same

**Comment by mcimadamore** [comment]
> > > Even with `arrayElementVarHandle` it's about the same
> > 
> > 
> > This is very odd, and I don't have a good explanation as to why that is the case. What does the baseline (confined arena) look like for `arrayElementVarHandle` ?
> 
> Pretty much exactly the same

So, that means that `arrayElementVarHandle` is ~4x faster than memory segment? Isn't that a bit odd?

**Comment by uschindler** [comment]
Thins looks all promising! Together with making sure that Apache Solr and Elasticsearch/Opensearch close indexes one-by-one in a separate thread (with the PR https://github.com/apache/lucene/pull/13570 in place, too), the issues should be fixed.

What is the issue with memory fences?

**Comment by JornVernee** [comment]
> > This is what I was thinking of as well. close() on a shared arena can be called by any thread, so it would be possible to have an executor service with 1-n threads that is dedicated to closing memory.
> 
> This delays both the closing of the Arena and the freeing of the segments, so bugs may be not discovered if the arena is accessed in between the time the thread pool is notified and the time the close() is effectively called.

Closing the arena is what requires the handshake, which is where the majority of the cost is. I don't see the point in closing synchronously, but then freeing the memory asynchronously, since the latter is relatively cheap.

**Comment by uschindler** [comment]
> > > This is what I was thinking of as well. close() on a shared arena can be called by any thread, so it would be possible to have an executor service with 1-n threads that is dedicated to closing memory.
> > 
> > 
> > This delays both the closing of the Arena and the freeing of the segments, so bugs may be not discovered if the arena is accessed in between the time the thread pool is notified and the time the close() is effectively called.
> 
> Closing the arena is what requires the handshake, which is where the majority of the cost is. I don't see the point in closing synchronously, but then freeing the memory asynchronously, since the latter is relatively cheap.

This is only a problem if you for example want to delete a mmapped file on Windows. This won't work as long as the memory is mmapped.

So there should be the option to allow async close() [if client supports it], but the defaulkt should be synchronized. I think this is what @forax suggested.

**Comment by mcimadamore** [comment]
> So, that means that `arrayElementVarHandle` is ~4x faster than memory segment? Isn't that a bit odd?

I did some more analyis of the benchmark. I first eliminated the closing thread, and started with two simple benchmarks:

```
@Benchmark
public int memorySegmentAccess() {
        int sum = 0;
        for (int i = 0; i < segment.byteSize(); i++) {
            sum += segment.get(JAVA_BYTE, i);
        }
        return sum;
    }
```

and

```
@Benchmark
public int otherAccess() {
        int sum = 0;
        for (int i = 0; i < array.length; i++) {
            sum += (byte)BYTE_HANDLE.get(array, i);
        }
        return sum;
    }
```

where the setup code is as follows:

```
static final int SIZE = 10_000;

    MemorySegment segment;
    byte[] array;

    static final VarHandle BYTE_HANDLE = MethodHandles.arrayElementVarHandle(byte[].class);

    @Setup
    public void setup() {
        array = new byte[SIZE];
        segment = MemorySegment.ofArray(array);
    }
```

With this, I obtained the following results:

```
Benchmark                            Mode  Cnt   Score   Error  Units
ConcurrentClose.memorySegmentAccess  avgt   10  13.879 ± 0.478  us/op
ConcurrentClose.otherAccess          avgt   10   2.256 ± 0.017  us/op
```

Ugh. It seems like C2 "blows up" at the third iteration:

```
# Run progress: 0.00% complete, ETA 00:05:00
# Fork: 1 of 1
# Warmup Iteration   1: 6.712 us/op
# Warmup Iteration   2: 5.756 us/op
# Warmup Iteration   3: 13.267 us/op
# Warmup Iteration   4: 13.267 us/op
# Warmup Iteration   5: 13.274 us/op
```

This might be a bug/regression. But, let's move on. I then tweaked the induction variable of the memory segment loop to be `long`, not `int` and I got:

```java
Benchmark                            Mode  Cnt  Score   Error  Units
ConcurrentClose.memorySegmentAccess  avgt   10  2.764 ± 0.016  us/op
ConcurrentClose.otherAccess          avgt   10  2.240 ± 0.016  us/op
```

Far more respectable! And now we have a good baseline, since both workloads take amount the same time, so we can use them to draw interesting comparisons. So, let's add back a thread that does a shared arena close:

```java
Benchmark                                        Mode  Cnt   Score   Error  Units
ConcurrentClose.sharedClose                      avgt   10  12.001 ± 0.061  us/op
ConcurrentClose.sharedClose:closing              avgt   10  19.281 ± 0.323  us/op
ConcurrentClose.sharedClose:memorySegmentAccess  avgt   10   9.802 ± 0.314  us/op
ConcurrentClose.sharedClose:otherAccess          avgt   10   6.921 ± 0.151  us/op
```

This is with vanilla JDK. If I apply the changes in this PR, I get this:

```java
Benchmark                                        Mode  Cnt   Score   Error  Units
ConcurrentClose.sharedClose                      avgt   10  10.837 ± 0.241  us/op
ConcurrentClose.sharedClose:closing              avgt   10  20.337 ± 1.674  us/op
ConcurrentClose.sharedClose:memorySegmentAccess  avgt   10   8.672 ± 0.993  us/op
ConcurrentClose.sharedClose:otherAccess          avgt   10   3.501 ± 0.162  us/op
```

This is good. Note how `otherAccess` improved almost 2x, as the code is no longer redundantly de-optimized. Now, we know that, even for memory segment access, we can avoid redundant deopt once JDK-8290892 is fixed. To simulate that, I've dropped the lines which apply the conservative deoptimization in `scopedMemoryAccess.cpp` and ran the bench again:

```
Benchmark                                        Mode  Cnt   Score   Error  Units
ConcurrentClose.sharedClose                      avgt   10   8.957 ± 0.089  us/op
ConcurrentClose.sharedClose:closing              avgt   10  18.898 ± 0.338  us/op
ConcurrentClose.sharedClose:memorySegmentAccess  avgt   10   4.403 ± 0.054  us/op
ConcurrentClose.sharedClose:otherAccess          avgt   10   3.571 ± 0.042  us/op
```

Ok, now both accessor threads seem faster.

If I swap the shared arena close with a confined arena close I get this:

```
Benchmark                                        Mode  Cnt   Score    Error  Units
ConcurrentClose.sharedClose                      avgt   10   1.760 ±  0.008  us/op
ConcurrentClose.sharedClose:closing              avgt   10  ≈ 10⁻³           us/op
ConcurrentClose.sharedClose:memorySegmentAccess  avgt   10   2.912 ±  0.016  us/op
ConcurrentClose.sharedClose:otherAccess          avgt   10   2.367 ±  0.009  us/op
```

Summing up:
* there is some issue involving segment access with `int` induction variable which we should investigate separately
* this PR significantly improves performance of threads that are not touching memory segments, even under heavy shared arena close loads
* performance of unrelated memory segment access is still affected by concurrent shared arena close. This is due to conservative deoptimization which will be removed once JDK-8290892 is fixed
* when all fixes will be applied, the performance of the accessing threads gets quite close to ideal, but not 100% there. The loss seems in the acceptable range - given that this benchmark is closing shared arenas in a loop, arguably the worst possible case.

**Comment by uschindler** [comment]
Thanks @mcimadamore, this sound great! I am so happy that we at least reduced the overhead for non-memory segment threads. This will also be the case for Lucene/Solr because we do not read from segments all the time, we also have other code sometimes executed between reads from memory segments :-)

So +1 to merge this and hopefully backport it at least to 21? This would be great, but as it is not a bug not strictly necessary.

We should open issues for the int problem and work on  JDK-8290892.

**Comment by mcimadamore** [comment]
> So +1 to merge this and hopefully backport it at least to 21?

Backport to 21 is difficult, given the handshake code there is different (and, FFM is preview there). But, might be more possible for 22. I have notified Roland of the `int` problem, will update once I know more about the nature of this issue.

**Comment by uschindler** [comment]
> > So +1 to merge this and hopefully backport it at least to 21?
> 
> Backport to 21 is difficult, given the handshake code there is different (and, FFM is preview there). But, might be more possible for 22. I have notified Roland re. the `int` problem, will update once I know more about the nature of this issue.

Ah I remember: the tristate! All fine.

**Comment by forax** [comment]
Even if the int vs long issue is fixed for this case, i think we should recommand to call `withInvokeExactBehavior()` after creating any VarHandle so all the auto-conversions are treated as runtime errors.

This is what i do with my students (when using compareAndSet) and it makes this kind of perf issue easy to find and easy to fix.

**Comment by mcimadamore** [comment]
> Even if the int vs long issue is fixed for this case, i think we should recommand to call `withInvokeExactBehavior()` after creating any VarHandle so all the auto-conversions are treated as runtime errors.
> 
> This is what i do with my students (when using compareAndSet) and it makes this kind of perf issue easy to find and easy to fix.

Note that this has nothing to do with implicit conversion, as the memory segment var handle is called by our implementation, with the correct type (a long). This is likely an issue with bound check elimination with "long loops".

**Comment by JornVernee** [comment]
Added JVMCI/Graal support, courtesy of @c-refice

/contributor add @c-refice

File: src/jdk.internal.vm.ci/share/classes/jdk/vm/ci/hotspot/HotSpotResolvedJavaMethod.java [hunk 4]
```diff
@@ -58,6 +58,13 @@ public interface HotSpotResolvedJavaMethod extends ResolvedJavaMethod {
      */
     boolean hasReservedStackAccess();
 
+    /**
+     * Returns true if this method has a {@code Scoped} annotation.
```
Comment (dougxc) [inline src/jdk.internal.vm.ci/share/classes/jdk/vm/ci/hotspot/HotSpotResolvedJavaMethod.java:62]: Can you please make this a qualified name: `jdk.internal.misc.ScopedMemoryAccess.Scoped`.
That makes it easier for someone not familiar with the code base to find.

Comment (JornVernee) [reply src/jdk.internal.vm.ci/share/classes/jdk/vm/ci/hotspot/HotSpotResolvedJavaMethod.java]: Done

**Comment by JornVernee** [comment]
As discussed offline, JVMCI/Graal changes will be handled by a followup PR.

**Comment by ChrisHegarty** [comment]
Thanks for the discussion and changes in this PR - it's super helpful ( in what we can do to workaround ), as well as a great improvement for the future.

**Comment by ChrisHegarty** [comment]
>  Effectively, once all the issues surrounding reachability fences will be addressed, we should be able to achieve numbers similar to above even in the case of shared close.

Is there an issue where I can follow this?

**Review by vnkozlov — APPROVED** [review]
I am fine with compiler and CI changes - it is just marking nmethod as having scoped access.

**Review by uschindler — APPROVED** [review]

**Review by iwanowww — APPROVED** [review]
Looks good.

**Comment by mcimadamore** [comment]
> * there is some issue involving segment access with `int` induction variable which we should investigate separately

This issue is tracked here:
https://bugs.openjdk.org/browse/JDK-8336759
